### PR TITLE
Add ModelContainer to romancal.datamodels.

### PR DESCRIPTION
**Pull request by mairanteodoro**
Resolves:
- [RCAL-568](https://jira.stsci.edu/browse/RCAL-568);
- [RCAL-468](https://jira.stsci.edu/browse/RCAL-468).

This PR addresses moving `ModelContainer` out of `roman_datamodels` and into `romancal` to follow the same structure that has been implemented for JWST, which takes care of [RCAL-568](https://jira.stsci.edu/browse/RCAL-568).

This version of `ModelContainer` is able to handle ASN files, which takes care of [RCAL-468](https://jira.stsci.edu/browse/RCAL-468). `ModelContainer` also supports slicing and dicing.

Previous unit tests for `ModelContainer` were updated and new ones were added.

Regression tests are passing:
https://plwishmaster.stsci.edu:8081/job/RT/job/Roman-Developers-Pull-Requests/223/

**Checklist**
- [X] added entry in `CHANGES.rst` under the corresponding subsection
- [X] updated relevant tests
- [X] updated relevant documentation
- [ ] updated relevant milestone(s)
- [X] added relevant label(s)
- [X] added regression test result

### CODE REVIEW

File: romancal/datamodels/tests/test_datamodels.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    assert image_model.test_attr == "test_attr_new_value"
+    # test ValueError
+    with pytest.raises(Exception) as e:
+        image_model["_test_attr"] = "test_attr_some_other_value"
+    assert e.type == ValueError
+
+
+def test_modelcontainer_init():
+    img = utils.mk_level1_science_raw()
+    m = rdm.ImageModel(img)
```
Comment (nden) [inline romancal/datamodels/tests/test_datamodels.py:77]: I am not sure why this works. Shouldn't it raise an error because a Level 1 file (i.e. 3D array) is being assigned to an ImageModel (i.e. 2D array)? Or am I misunderstanding it?

**Review by nden — CHANGES_REQUESTED** [review]

File: romancal/datamodels/container.py [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+        self._models.pop(index)
+
+    def copy(self, memo=None):
+        """
+        Returns a deep copy of the models in this model container.
+        """
+        return copy.deepcopy(self, memo=memo)
+
+    def close(self):
+        if not self._iscopy and self._asdf is not None:
```
Comment (nden) [inline romancal/datamodels/container.py:253]: Does a Roman `ModelContainer` have a `_asdf` attribute? What is `self._asdf is not None` checking?

File: romancal/datamodels/container.py [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+        for m in self._models:
+            if isinstance(m, rdm.DataModel):
+                result.append(m.copy())
+            else:
+                result.append(m)
+        return result
+
+    def close(self):
+        if not self._iscopy and self._asdf is not None:
+            self._asdf.close()
```
[the remark this answers is not in the document]
Comment (nden) [reply romancal/datamodels/container.py:254]: I don't think this comment is resolved. I am not sure what `self._asdf.close()` means.

File: romancal/datamodels/container.py [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+    def from_asn(self, asn_data, asn_file_path=None):
+        """
+        Load fits files from a Roman association file.
+
+        Parameters
+        ----------
+        asn_data
+            A ``roman_datamodels.associations.Association`` association dictionary
+
+        asn_file_path: str
```
[the remark this answers is not in the document]
Comment (nden) [reply romancal/datamodels/container.py:283]: Not resolved: The lines need to be
`asn_file_path : str`

File: romancal/datamodels/container.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+                filepath = op.join(asn_dir, member["expname"])
+                update_model = any(attr in member for attr in RECOGNIZED_MEMBER_FIELDS)
+                if update_model or self._save_open:
+                    m = rdm.open(filepath, memmap=self._memmap)
+                    m.meta["asn"] = {"exptype": member["exptype"]}
+                    for attr, val in member.items():
+                        if attr in RECOGNIZED_MEMBER_FIELDS:
+                            if attr == "tweakreg_catalog":
+                                val = op.join(asn_dir, val) if val.strip() else None
+                            setattr(m.meta, attr, val)
```
Comment (nden) [inline romancal/datamodels/container.py:319]: A test covering lines 317-320 i needed. Isn't this what tweakreg is going to use? Let's make sure it works.

File: romancal/datamodels/container.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nden) [reply romancal/datamodels/container.py:254]: Which means the `close` method is not tested.

File: romancal/datamodels/container.py [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+            if group_id in group_dict:
+                group_dict[group_id].append(model)
+            else:
+                group_dict[group_id] = [model]
+
+        return group_dict.values()
+
+    @property
+    def to_association(self):
+        pass
```
Comment (nden) [inline romancal/datamodels/container.py:440]: Why is this needed in this form?

File: romancal/datamodels/container.py [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+
+        Returns
+        -------
+        dict
+        """
+        crds_header = {}
+        if len(self._models):
+            model = self._models[0]
+            model = model if isinstance(model, rdm.DataModel) else rdm.open(model)
+            crds_header |= model.get_crds_parameters()
```
Comment (nden) [inline romancal/datamodels/container.py:484]: I think this will work, I'm just not sure why it is necessary when a model is guaranteed to have a `get_crds_parameters` method which returns a valid dictionary. Perhaps it's best to avoid the extra computation, just return the output of that method.

**Comment by nden** [comment]
I think it would be useful also to have some example association files from @ddavis-stsci and use them in the unit tests. These are small files and can go in the repo.

**Comment by ddavis-stsci** [comment]
You can find some example associations at 
/grp/roman/ddavis/roman/data/associations/

File: romancal/datamodels/container.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mairanteodoro) [reply romancal/datamodels/container.py:254]: We added a unit test for `.close` now.

File: romancal/datamodels/container.py (hunk 6 above)
Comment (mairanteodoro) [reply romancal/datamodels/container.py]: Removed for now.

**Review by nden — APPROVED** [review]
The changes now look good to me.
I'll restart the tests to see if this was a Github fluke.

**Comment by mairanteodoro** [comment]
@zacharyburnett is looking into it.

**Comment by nden** [comment]
spacetelescope/roman_datamodels#203 should fix the current test failure
